osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -1,181 +1,317 @@
+    @Override
+    public byte[] getKey(String deviceIdentification, SecurityKeyType keyType) {
+        LOGGER.info("Retrieving {} for device {}", keyType.name(), deviceIdentification);
+        return this.getKeys(deviceIdentification, Arrays.asList(keyType)).get(keyType);
+    }
 
-            return decryptedKey;
+    @Override
+    public Map<SecurityKeyType, byte[]> getKeys(String deviceIdentification, List<SecurityKeyType> keyTypes) {
+        GetSecretsRequest request = this.createGetSecretsRequest(deviceIdentification, keyTypes);
+        GetSecretsResponse response = this.secretManagementClient.getSecretsRequest(request);
+        this.validateGetResponse(keyTypes, response);
+        return this.convertSoapSecretsToSecretMapByType(response.getTypedSecrets().getTypedSecret());
+    }
 
-        } catch (Exception e) {
-            LOGGER.error("Error while retrieving encryption key", e);
+    private void validateGetResponse(List<SecurityKeyType> keyTypes, GetSecretsResponse response) {
+        if (!OsgpResultType.OK.equals(response.getResult()) || response.getTypedSecrets() == null
+                || response.getTypedSecrets().getTypedSecret() == null) {
+            throw new IllegalStateException(
+                    "Invalid/incomplete soap response: resulttype=" + response.getResult().value());
+        }
+        if (keyTypes.size() != response.getTypedSecrets().getTypedSecret().size()) {
+            throw new IllegalStateException("Unexpected number of secrets in response");
         }
-        return new byte[0];
     }
 
-    @Override
-    public byte[] getMbusDefaultKey(String mbusDeviceIdentification) {
-        throw new NotImplementedException();
+    private Map<SecurityKeyType, byte[]> convertSoapSecretsToSecretMapByType(List<TypedSecret> soapSecrets) {
+        Function<TypedSecret, SecurityKeyType> convertType = ts -> SecurityKeyType.fromSecretType(ts.getType());
+        Function<TypedSecret, byte[]> convertSecret = ts -> this.decryptSoapSecret(ts);
+        return soapSecrets.stream().collect(Collectors.toMap(convertType, convertSecret));
     }
 
-    @Override
-    public byte[] getMbusUserKey(String mbusDeviceIdentification) {
-        throw new NotImplementedException();
+    private GetSecretsRequest createGetSecretsRequest(String deviceIdentification, List<SecurityKeyType> keyTypes) {
+        GetSecretsRequest request = new GetSecretsRequest();
+        request.setDeviceId(deviceIdentification);
+        request.setSecretTypes(new SecretTypes());
+        List<SecretType> secretTypeList = request.getSecretTypes().getSecretType();
+        keyTypes.stream().forEach(kt -> secretTypeList.add(kt.toSecretType()));
+        return request;
     }
 
+    /*@Override
+    public void aesDecryptAndStoreNewKey(String deviceIdentification, SecurityKeyType keyType, byte[] encryptedKey)
+            throws FunctionalException {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') This block of commented-out lines of code should be removed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ACommentedOutCodeLine)